core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java

```diff
@@ -50,13 +42,64 @@ public long nullValueCount() {
     throw new IllegalStateException("Shouldn't access this method, as this metric is tracked in file statistics. ");
   }
 
-  @Override
-  public ByteBuffer lowerBound() {
-    throw new IllegalStateException("Shouldn't access this method, as this metric is tracked in file statistics. ");
+  public static class FloatFieldMetricsContext extends AbstractFloatFieldMetricsContext<Float> {
+    public FloatFieldMetricsContext(int id) {
+      super(id);
+    }
+
+    @Override
+    public void updateMetricsContext(Float value) {
+      if (Float.isNaN(value)) {
+        this.nanValueCount++;
+      } else {
+        if (lowerBound == null || Float.compare(value, lowerBound) < 0) {
+          this.lowerBound = value;
+        }
+        if (upperBound == null || Float.compare(value, upperBound) > 0) {
+          this.upperBound = value;
+        }
+      }
+    }
   }
 
-  @Override
-  public ByteBuffer upperBound() {
-    throw new IllegalStateException("Shouldn't access this method, as this metric is tracked in file statistics. ");
+  public static class DoubleFieldMetricsContext extends AbstractFloatFieldMetricsContext<Double> {
```

Comment:
This class is suspiciously similar to the builder pattern, so I would probably just convert it to be a builder for `FieldMetrics` instead:

* Rename `DoubleFieldMetricsContext` to `DoubleFieldMetrics.Builder` and create it with `builderFor(int id)`
* Rename `updateMetricsContext` to something shorter, like `addValue`
* Rename `buildMetrics` to just `build`